### DISCUSSION TITLE
Date formats on Dashboards

### DISCUSSION THREAD

**Discussion by Ed-M72** [teslamate-org/teslamate#1512] [Ideas]
@DrMichael.  I know you have been working own the dashboards, I could do with some advice. There are 8 dashboards with date formats, I would like to standardise the date formats across them.

Below is a table showing the eight dashboards and their output when displayed in a browser with a timezone of Europe/London.  As you can see the date formats are inconsistent, arising from either a restriction in Grafana, the use of the distance unit as a flag for the date format and a mix between DateTime ISO & DateTime local being defined in the dashboards.

There is a restriction on the Grafana table which means you can only use either an US or ISO date format, which will stop me adding the date formats for other countries, e.g. the U.K.

I think the only way to make all the dashboards consistent, and readable to everyone is to add a new setting called "Date Formats", with the options "American/European" or "MM/DD/YYYY, YYYY-MM-DD", and use this to consistently set the date format across the existing and any new dashboards.

I have added an extra "date format" dropdown on the settings page of a local dev build, but would like some feedback before I spend time updating all the dashboards.

Below is the table I created during the initial analysis.
Ed

@adriankumpf 

  | Dashboard | Current Display (Miles) | Current Display (Kilometres) | Current Format | Proposal
-- | -- | -- | -- | -- | --
1 | Charges | MM/DD/YYYY | YYYY-MM-DD | Manual Overide based on Distance unit | Manual Override based on Date Format setting (American/European).
2 | Drives | MM/DD/YYYY | YYYY-MM-DD | Manual Overide based on Distance unit | Manual Override based on Date Format setting (American/European).
3 | Locations | YYYY-MM-DD | YYYY-MM-DD | Date with Datetime local | Manual Override on name_[Date_format]. Options become Datetime US or Datetime ISO.
4 | States | YYYY-MM-DD | YYYY-MM-DD | Straight from DB, no overrides | Manual Override on name_[Date_format]. Options become Datetime US or Datetime ISO.
5 | Statistics | YYYY-MM-DD | YYYY-MM-DD | Straight from DB, no overrides | Manual Override on name_[Date_format]. Options become Datetime US or Datetime ISO.
6 | Updates | YYYY-MM-DD | YYYY-MM-DD | Datetime ISO | Manual Override on name_[Date_format]. Options become Datetime US or Datetime ISO.
7 | Vampire Drain | MM/DD/YYYY | YYYY-MM-DD | Manual Overide based on Distance unit | Manual Override based on Date Format setting (American/European).
8 | Dutch Tax | DD/MM/YYYY | DD/MM/YYYY | Date with Datetime local | Manual Override on name_[Date_format]. Options become Datetime US or Datetime ISO.

**Comment by DrMichael** [comment]
Hi, well, we had a discussion about that before: https://github.com/adriankumpf/teslamate/discussions/1320

I tried to tinker with
```
#      - GF_USE_BROWSER_LOCALE=true
#      - GF_DATE_FORMATS_INTERVAL_HOUR=DD/MM HH:mm
#      - GF_DATE_FORMATS_INTERVAL_DAY=DD/MM
#      - GF_DATE_FORMATS_FULL_DATE=DD.MMM.YYYY hh:mm:ss
```
in the docker-compose.yml, but was not really successful. Might be, that Grafana does not really support that well. But I think, that would be the way to go. And I agree, it should not be based on km/mi setting.

**Reply by Ed-M72** [reply]
Yes, you are correct, I tried the the docker environment variables first, they seem to control the main dashboard date restrictions, however the environment variables make absolutely no difference to the plugins, which is where you really need them.

**Comment by DrMichael** [comment]
Ah, ok. I really like the idea of harmonising the date formats. I personally could live with ISO format `ISO 8601/SQL standard | 1997-12-17 07:37:16-08` although it is not the date format of my native locale, de.

Formats like MM/DD/YYYY or DD.MM.YYY could lead to difficulties, when sorting...

**Reply by Ed-M72** [reply]
Agreed. I could live with that format too.  Perhaps we standardise all of them on ISO 8601?  That format is understandable by everybody, and reduces the amount of effort to maintain the dashboards.

I discovered last night that the Stats panel only accepts one hard-coded date format so even with a user defined date format in the config it would always display as ISO 8601 regardless of a user selecting MM/DD/YYYY.

**Comment by DrMichael** [comment]
Well, one could use `Datetime ISO` also in Statistics, but that includes the also seconds, which is not needed/wanted.

**Comment by rossnick** [comment]
I just came across teslamate. Wow !

I was searching on how to change the datetime formats in the grafana dashboards, and came to this discussion.

Any hints on how to achieve this ?

**Reply by DrMichael** [reply]
This here is quite outdated...
The date is formatted according to your browser settings. You would need to change it there.

**Reply by rossnick** [reply]
Oh, really ! I didn't know browsers could do that. I set my firefox to use the system settings to format time. But still, it the grafana dashboard shows time in 12H format and the system is showing 24...

**Reply by 265chemic** [reply]
> This here is quite outdated... The date is formatted according to your browser settings. You would need to change it there.

Thanks for this. Was driving me nuts playing with env variables in docker-compose (which does not work).  Changing my default browser locale did the trick - thanks!!!
